### PR TITLE
[DF] Add first tutorial for distributed rdf

### PR DESCRIPTION
**Pull request by vepadulano**

### CODE REVIEW

**Comment by vepadulano** [comment]
@phsft-bot build just on mac11.0/cxx17

File: tutorials/dataframe/distrdf001_vecOpsHEP.py
Comment (eguiraud) [inline tutorials/dataframe/distrdf001_vecOpsHEP.py:57]: let's keep tutorials focused -- you can just point to the other tutorials rather than including non-distributed ways to do the same thing here

File: tutorials/dataframe/distrdf001_vecOpsHEP.py
Comment (eguiraud) [inline tutorials/dataframe/distrdf001_vecOpsHEP.py:1]: `distrdf001_vecOpsHEP` is a "scary" title for the very first tutorial. As per the PPP yesterday, we learned that tutorial number 1 should be a simple hand-holding introduction to the topic (in this case, distRDF).

I think it's ok to use RVecs (we have to assume that users looking at distrdf stuff are familiar with "standard" RDF content) but I would not highlight it so much

File: tutorials/dataframe/distrdf001_vecOpsHEP.py
Comment (eguiraud) [inline tutorials/dataframe/distrdf001_vecOpsHEP.py:65]: I don't understand this explanation of the default behavior :sweat_smile:"the computations are run through the distributed Spark backend" clashes with "run the computations locally". I would change the last 3 sentences to something direct like "By default, a `Spark.RDataFrame` runs on N/all/1 _local_ cores using Spark to distribute the work among processors."

File: tutorials/dataframe/distrdf002_spark_connection.py
Comment (eguiraud) [inline tutorials/dataframe/distrdf002_spark_connection.py:8]:
```suggestion
## SparkContext object created with the desired options. After this initial
```

File: tutorials/dataframe/distrdf001_vecOpsHEP.py
Comment (eguiraud) [inline tutorials/dataframe/distrdf001_vecOpsHEP.py:31]: is there no way to automate this?

**Review by eguiraud — COMMENTED** [review]
some discussion needed before approval

File: tutorials/dataframe/distrdf001_vecOpsHEP.py
Comment (vepadulano) [reply tutorials/dataframe/distrdf001_vecOpsHEP.py:57]: I imagined this would be the point of this tutorial, but if you think it's not worth it then I would remove it altogether and make another one that doesn't reference `df017`

File: tutorials/dataframe/distrdf001_vecOpsHEP.py
Comment (vepadulano) [reply tutorials/dataframe/distrdf001_vecOpsHEP.py:1]: Ok fair enough, then what's currently called  `distrdf002` represents more of an introduction, e.g. creating a dataset from scratch and plotting a couple of histograms. The only extra step shown there is the connection to a spark cluster which I would label as "more advanced", so I'm a bit in doubt what the introduction could be

File: tutorials/dataframe/distrdf001_vecOpsHEP.py
Comment (vepadulano) [reply tutorials/dataframe/distrdf001_vecOpsHEP.py:65]: Ok thank you, will rephrase

File: tutorials/dataframe/distrdf001_vecOpsHEP.py
Comment (vepadulano) [reply tutorials/dataframe/distrdf001_vecOpsHEP.py:31]: We could *probably* pass environment variables to CTest when running these tutorials? Although I thought it was also worth showing it here because it's something that can happen to any spark user

Comment (eguiraud) [reply tutorials/dataframe/distrdf001_vecOpsHEP.py:31]: yes if users _have_ to set this otherwise things might break we have to show it. the question was whether we can set this for users in general

File: tutorials/dataframe/distrdf001_vecOpsHEP.py
Comment (eguiraud) [reply tutorials/dataframe/distrdf001_vecOpsHEP.py:1]: why would you use distRDF if you don't want to connect to a spark cluster though? also that's basically the only difference w.r.t. "normal" RDF usage, so that's what a distRDF tutorial should highlight -- for explanations/tutorials on existing RDF features you can/should point to existing docs/tutorials

File: tutorials/dataframe/distrdf001_vecOpsHEP.py
Comment (vepadulano) [reply tutorials/dataframe/distrdf001_vecOpsHEP.py:31]: It's complicated. The only real general rule is: "the environment of the distributed executors (i.e. Python processes running pyspark map/reduce functions on the distributed computing nodes) should have all dependencies needed to run the map/reduce functions".

Now, in *this* particular case. Since the tutorials have to run in the user laptop (or locally in our CI node for that matter), our best guess is that the environment where the Python tutorial is started is also the desired environment for the PySpark executors to run in, hence setting `PYSPARK_PYTHON` to the current python executable makes sure the environments are aligned. A very common error is to setup a Python virtual environment, install pyspark there, run a pyspark application inside the virtual environment, then the pyspark processes responsible of running map/reduce functions will pick the default system python which might be a different version or not have all the required dependencies and fail.

This is also a very common situation in a distributed environment for similar reason. User starts a pyspark application in their laptop / web interface representing their personal session (e.g. SWAN), then the distributed workers pick by default whatever's the Python executable on their system (which in principle could be very heterogeneous). 

Setting `PYSPARK_PYTHON` can be done in multiple ways:
1. Like done here, setting the environment variable in the same Python session of the Spark driver
2. Setting this environment variable in $SPARK_HOME/conf/spark-env.sh
3. Setting it as part of the parameters to a SparkConf object, e.g.
   ```python
   import pyspark
   conf = pyspark.SparkConf().set("spark.pyspark.python", "/path/to/python")
   sc = pyspark.SparkContext(conf=conf)
   ```

This has to be taken into account for better or worse, either by the sysadmin of the Spark cluster or by the users themselves (especially if they want to provide their default custom environment). If we want to take SWAN as an example, a user has to click on the "connect to cluster" button in the jupyter notebook interface in their SWAN session, then they can add any Spark specific parameter in the configuration panel, then SWAN internally takes care of creating a `SparkConf` object that will also make sure the Python environment on the workers is aligned with the Python environment on the driver.

Comment (vepadulano) [reply tutorials/dataframe/distrdf001_vecOpsHEP.py:31]: Now to go back to our tutorials, also following your comment above

> why would you use distRDF if you don't want to connect to a spark cluster though?

I take that it's ok if we show the creation of `SparkConf` and `SparkContext` objects as part of any `distrdf` tutorial? In that case, I think we can also write the `spark.pyspark.python` variable in there

Comment (vepadulano) [reply tutorials/dataframe/distrdf001_vecOpsHEP.py:31]: Note for the future: this kind of issue will present itself in any other distributed backend so we'll have to adapt to how the various frameworks manage aligning environments between the submitting application and the distributed workers

Comment (vepadulano) [reply tutorials/dataframe/distrdf001_vecOpsHEP.py:31]: Another note on pyspark functionality: it's not compulsory to have *exactly* the same environment between driver and workers, that's why it's not enforced by Spark itself. The only requirement is to have the same Python version.

Comment (vepadulano) [reply tutorials/dataframe/distrdf001_vecOpsHEP.py:31]: TL;DR

No, we cannot set it in general for all users, actually it would go against a Spark feature.
Yes, if the worker environment doesn't have the same Python version and dependencies expected by the driver environment the pyspark application will crash. Thus, this setting is needed somewhere (see the bullet points above for possible options)

File: tutorials/dataframe/distrdf001_vecOpsHEP.py
Comment (vepadulano) [reply tutorials/dataframe/distrdf001_vecOpsHEP.py]: Tutorial has been deleted in the latest commit

File: tutorials/dataframe/distrdf001_vecOpsHEP.py
Comment (vepadulano) [reply tutorials/dataframe/distrdf001_vecOpsHEP.py]: Right, the latest commit includes the previously called `distrdf002` tutorial as the only tutorial, which is more verbose in the explanation of the connection to a Spark cluster while also keeping the application very simple

**Comment by vepadulano** [comment]
It appears that the configuration option `pyspark.spark.python` is somehow ineffective and Spark really only cares about the environment variable `PYSPARK_PYTHON` https://issues.apache.org/jira/browse/SPARK-26404 . This is a bummer...

**Comment by vepadulano** [comment]
@phsft-bot build just on mac11.0/cxx17

File: tutorials/dataframe/distrdf001_spark_connection.py
Comment (stwunsch) [inline tutorials/dataframe/distrdf001_spark_connection.py:64]: The Spark specifics look a little bit scary here. Why not compressing it with some nice Python syntax? I see that you want to explain in detail how to setup the Spark side, but probably it would be worth to put things together:

```python
sparkconf = pyspark.SparkConf().setAll({
         "spark.app.name": "distrdf001_spark_connection",
         "spark.master": "local[4]"
        }.items())
sc = pyspark.SparkContext(conf=sparkconf)
```

Comment (stwunsch) [reply tutorials/dataframe/distrdf001_spark_connection.py:64]: And probably I would name `sc` just `sparkcontext`. Spelling this out makes it very clear what it is, without remembering the name of the class used for the ctor.

File: tutorials/dataframe/distrdf001_spark_connection.py
Comment (stwunsch) [reply tutorials/dataframe/distrdf001_spark_connection.py:60]: Oh wow you had a lengthy discussion above about this :D In the local case, the Python version should be automatically correct, right? So I'd probably just remove setting the pyspark python version.

Comment (vepadulano) [reply tutorials/dataframe/distrdf001_spark_connection.py:60]: Yes in the commit right before this one :( and I found out this very nasty jira issue https://issues.apache.org/jira/browse/SPARK-26404

**Comment by vepadulano** [comment]
The latest commit fixes the PYSPARK_PYTHON issue by setting the environment variable in the ctest invocation, tutorial is green on mac11 (the only label that runs tutorial by default and has pyspark installed at the same time) https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/112683/testReport/projectroot/runtutorials/tutorial_dataframe_distrdf001_spark_connection_py/  . I just noticed I need to change this line and this is good to go

```suggestion
- h_poisson = df_1.Histo1D(("poisson", "Exponential distribution", 50, 0, 30), "poisson")
+ h_poisson = df_1.Histo1D(("exponential", "Exponential distribution", 50, 0, 30), "exponential")
```

**Review by stwunsch — APPROVED** [review]
Perfect! :)

**Pull request merged**
